lib/puppetserver/ca/certificate_authority.rb

```diff
@@ -0,0 +1,210 @@
+require 'puppetserver/ca/utils/http_client'
+
+require 'json'
+
+module Puppetserver
+  module Ca
+    class CertificateAuthority
+
+      include Puppetserver::Ca::Utils
+
+      REVOKE_BODY = JSON.dump({ desired_state: 'revoked' })
+      SIGN_BODY =   JSON.dump({ desired_state: 'signed' })
+
+      def initialize(logger, settings)
+        @logger = logger
+        @client = HttpClient.new(settings)
+        @ca_server = settings[:ca_server]
+        @ca_port = settings[:ca_port]
+      end
+
+      # Returns a URI-like wrapper around CA specific urls
+      def make_ca_url(resource_type = nil, certname = nil)
+        HttpClient::URL.new('https', @ca_server, @ca_port, 'puppet-ca', 'v1', resource_type, certname)
+      end
+
+      def sign_certs(certnames)
+        put('certificate_status', certnames, SIGN_BODY, :sign)
+      end
+
+      def revoke_certs(certnames)
+        put('certificate_status', certnames, REVOKE_BODY, :revoke)
+      end
+
+      # Should this csr be in JSON? instead of string?
+      def submit_certificate_request(certname, csr)
+        put('certificate_request', [certname], csr, :submit)
+      end
+
+      # Make an HTTP PUT request to CA
+      # @param resource_type [String] the resource type of url
+      # @param certnames [Array] array of certnames
+      # @param body [JSON/String] body of the put request
+      # @param type [Symbol] type of error processing to perform on result
+      # @return [Boolean] whether all requests were successful
+      def put(resource_type, certnames, body, type)
+        url = make_ca_url(resource_type)
+        results = @client.with_connection(url) do |connection|
+          certnames.map do |certname|
+            url.resource_name = certname
+            result = connection.put(body, url)
+            process_results(type, certname, result)
+          end
+        end
+
+        results.all?
+      end
+
+      # logs the action and returns true/false for success
+      def process_results(type, certname, result)
+        case type
+        when :sign
+          case result.code
+          when '204'
+            @logger.inform "Successfully signed certificate request for #{certname}"
+            return true
+          else
```

Comment:
Is there any result that can't return 404? We'd also want it to be an error in most cases as well, right?

Reply:
^ Sorry Justin.  I don't undertand.  If its not 204, it will log the error.

Reply:
My bad, I was being obtuse. I meant that if 404 was a valid error state for this endpoint (which I think it's a valid response from the server for all endpoints), then the existing 404 handling would give the user a better error message then this more general one, and potentially allow us to fold this in with other result processing.

Reply:
Though if we have to duplicate all of our success and error messages combining them may not be much of a win.

Reply:
Lemme take a look about adding it back.

Reply:
ok.  I remember now
it was:
```
  when '404'
              @logger.err 'Error:'
              @logger.err "    Could not find certificate for #{certname}"
              success = false
```
But if there are SANs we also get a 404, and it that case we want the custom error message that the endpoint sends us and not something like "could not find certificate".  Thats why i changed it.

Reply:
added back.